### PR TITLE
Moved Corrections Report into Component

### PR DESCRIPTION
**Pull request by abrieff**
Addresses #124 
There are no real code changes here, other than:
1. Creates new component folder, and a new corrections_report.py file to encapsulate all corrections related code
2. Moves all functions in the corrections report out of their nesting, which is easier for testing, ipython, etc.

Some decisions we might answer here, or in a future PR:
1. How should data ingestion work?
2. How should importing components work?
3. How should importing libraries work? Is there any way to avoid doing that for every file?
4. Should a component be a class instead of just a group of functions?

Figured I'd get the ball rolling though with minimal extra changes. Maybe 2 🐰 reviewers to approve of the architectural change

### CODE REVIEW

**Comment by SashaWeinstein** [comment]
This is great, thanks for getting the ball rolling on this improvement. Here are my thoughts on your questions
1. I everything in the pluto helpers but `get_data` and `blacklist_branches` can be moved to an app-wide `load.py`, that upgrade should be part of the refactor we are starting here
2. I think the way we import this component is a fine approach going forward, do you see something wrong with it?
3. I think we need to import in every file, don't know a better way in python
4. I think a class could be helpful to make the code more readable. The readability issue I see is that it's not clear in the component file which function is responsible for returning the entire component. I think a class could help make that hierarchy clear. Additionally in my mental model I think of a component of a thing that exists, which is better captured as an object than a function which goes off once but doesn't persist.

**Comment by abrieff** [comment]
1. So you think I should include the digital ocean abstracting stuff as part of this PR? can do that
2. Not really - I toyed with how we might include all components as a whole to avoid having to alter the imports every time we add a component - making each one a class would probably make that easier (didn't want to include a bunch of private functions unnecessarily) (like import src.pluto.components as components, components.corrections_report(), etc.)
3. 👍 
4. That makes sense to me, but will take some thinking to make it jive with how streamlit behaves (if st commands are run again, they'll just print more stuff)

**Comment by SashaWeinstein** [comment]
Sorry I wasn't clear in 1, I meant the DO abstraction should be part of this round of PR's, but not part of this PR specifically

**Comment by SashaWeinstein** [comment]
Yea I was wondering the same thing. Maybe the class can have a `__call__` that returns the associated streamlit elements?

**Comment by abrieff** [comment]
What is a __call__?

**Comment by abrieff** [comment]
My usual instinct with classes like this is to use duck typing, so that all of these components define the same public method (so you could like, iterate on them in a list or use them without knowing how the innards work.) So each component has like, one public "display" method or something like that that prints the component

**Comment by SashaWeinstein** [comment]
I was talking about [https://www.geeksforgeeks.org/__call__-in-python/] which is implemented in [zap open data](https://github.com/NYCPlanning/db-zap-opendata/blob/822f26aea4ff9dd0ebb435cfb32c86352db5f31f/src/runner.py#L171) in a way I like. 

I imagined we would just instantiate the component object and have it display it's content, which just requires an init to make the streamlit elements and a call to return them. Do we need more steps than that? If so then I think duck typing is the right pattern, each component class should be able to do the same things

**Comment by abrieff** [comment]
Neat, that makes sense. Think its ultimately the same basic concept just a little cleaner

**Comment by SashaWeinstein** [comment]
Call is cleaner? If you can think of more than just one function we need I'm open to duck typing

**Comment by SashaWeinstein** [comment]
I can implement the class this afternoon if you like

**Comment by abrieff** [comment]
yeah I think call is cleaner - and go for it

**Comment by SashaWeinstein** [comment]
@abrieff I refactored the report itself and the two sections into classes. Think this helps the reader discern the components and sub-components from the helper functions. Let me know what you think

File: src/pluto/components/corrections_report.py
Comment (abrieff) [inline src/pluto/components/corrections_report.py:59]: The only thing I wonder here is if this is a good opportunity for class inheritance - ie: making a base CorrectionsSection class that includes the __init__ and all of the floating functions below, then these classes look just like:
```
  class AppliedCorrectionsSection(CorrectionsSection):
    def __call__(self):
       # stuff
```

Comment (SashaWeinstein) [reply src/pluto/components/corrections_report.py:59]: Before I get into the code here I wanted to share my thoughts on terminology. The terminology that seemed best to me while I was working on this yesterday is that each data product has a page, pages have reports (the corrections section is a report) and reports have sections. Our refactor here is bringing a report into a component file. Does this terminology seem right to you? Now is the time to think about it, future refactors into component files should follow whatever patterns we set here. 

In terms of this specific code, that inheritance would mean we don't have to pass `applied_corrections` or `not_applied_corrections` to the init of the section, we just reference `self.<appropriate corrections attr>` right? That would make the code neater, I'll make that change now

**Comment by SashaWeinstein** [comment]
Ok @abrieff take a look at this round of changes

File: src/pluto/components/corrections_report.py
Comment (abrieff) [inline src/pluto/components/corrections_report.py:7]: What does this do?

File: src/pluto/components/corrections_report.py
Comment (abrieff) [inline src/pluto/components/corrections_report.py:104]: Think this function can be deleted

**Review by abrieff — COMMENTED** [review]
lgtm after deleting the extra init! (cant approve since its my PR but when you're ready would say approve and we can merge)

File: src/pluto/components/corrections_report.py
Comment (SashaWeinstein) [reply src/pluto/components/corrections_report.py:104]: nice catch, will address in next commit

File: src/pluto/components/corrections_report.py
Comment (SashaWeinstein) [reply src/pluto/components/corrections_report.py:7]: it stand for "abstract base class", which is a base class that can't be instantiated. The only objects created with the implemented functionality are from classes that inherit from the abc.

Comment (abrieff) [reply src/pluto/components/corrections_report.py:7]: 👍 makes sense

**Comment by SashaWeinstein** [comment]
There are now two `__init__` calls, one for the report and one for the base class of the function 🔮 . Think we need a third reviewer as we've both contributed.

**Comment by abrieff** [comment]
right, that makes sense 👍

**Comment by SashaWeinstein** [comment]
I wrote out my thoughts in issue #148

**Review by td928 — APPROVED** [review]
don't have a lot to say except: Bravo! I think the new paradigm improve a lot on readability and create the hierarchy within the different parts that were hidden before. Really great learning wise reading through the thread and works on my end. 

I have one nit picking comment which is naming of "section". I think I am a little bothered by the vagueness with it. My alternative is kind of dumb which is "component". But I do think brining it back full circles with the folder structures in this case make some sense.

**Comment by SashaWeinstein** [comment]
Great! 
Very happy to talk about terminology here. I agree that section is vague. I think "component" is nice as it's clearly smaller, but I know it also has a technical meaning within front-end development like in [this blog post](https://dev.to/xavortm/what-are-components-in-the-front-end-and-why-do-we-need-them-2o2p) and also has a specific meaning in [streamlit world](https://streamlit.io/components).  This doesn't mean we can't use it though.

**Comment by td928** [comment]
Got it. "Component" does smell kind of javascript-y almost. On second thought, "section" does kind of flow with "report" so I am liking it a little more than before.

**Comment by abrieff** [comment]
we could rename the folder `reports` if that feels better

**Comment by td928** [comment]
> we could rename the folder `reports` if that feels better

for some reason I really don't like that lol

**Comment by abrieff** [comment]
haha okay, well we could leave it as is and you all could iterate on it in the future if you come up with something better?

**Comment by abrieff** [comment]
in my mind, section was kind of akin to the 
<section> html tag

**Pull request merged**
